LoopKit/InsulinKit/DoseStore.swift

```diff
@@ -1646,49 +1627,121 @@ extension DoseStore {
             return
         }
 
-        completion(.success(queryAnchor, queryResult))
-    }
+        // Determine the potential schedule basal end date used to query the insulin delivery store. Find the date of the
+        // latest pump event that can end a scheduled basal, namely either a resume or temp basal event.
+        let scheduledBasalEndDate = (pumpEvents.count == limit) ? pumpEvents.filter({ $0.delineatesScheduledBasal }).last?.endDate : nil
 
-    public func executePumpEventQuery(fromQueryAnchor queryAnchor: QueryAnchor?, limit: Int, completion: @escaping (PumpEventQueryResult) -> Void) {
-        var queryAnchor = queryAnchor ?? QueryAnchor()
-        var queryResult = [PersistedPumpEvent]()
-        var queryError: Error?
-
-        guard limit > 0 else {
-            completion(.success(queryAnchor, []))
-            return
-        }
+        // Get the doses within the pump event window
+        insulinDeliveryStore.getDoseEntries(start: queryAnchor.scheduledBasalStartDate, end: scheduledBasalEndDate, limit: limit, inclusive: false) { result in
+            switch result {
+            case .failure(let error):
+                completion(.failure(error))
+            case .success(let doses):
+                var doseEvents = doses.compactMap { $0.syncPumpEventWithModificationCounter }
+                doseEvents.sort(by: { SyncPumpEventWithModificationCounter.sort($0, $1) })
+
+                var events: [SyncPumpEventWithModificationCounter] = []
+
+                // Walk the two sets of events and reconcile. As the events are sorted by start date, the event with the earlier
```

Comment:
[nit] Consider splitting this logic out into separate function[s].  As it is, it is kinda hard to follow and get the big picture, all inside this one `executePumpQuery` function. The comments are helpful, though, thanks.